### PR TITLE
Version Scheduling Goals and Conditions

### LINKED ISSUE #1277: Store Shared Versions of Scheduling Goals and Conditions
Scheduling Goals and Conditions need to be versioned and those versions need to be reusable between plans.

### PR DESCRIPTION
**Pull request by Mythicaeda**
* **Tickets addressed:** Closes #1277
* **Review:** By commit  
* **Merge strategy:** Merge (no squash)  

## Description

Sister PR to #1278. 

Overview of changes:
- The Scheduling Conditions table has been split into a metadata and definitions table. **BREAKING CHANGE**
- Likewise, the Scheduling Goals table has been split into a metadata and definitions table. **BREAKING CHANGE**
- Definitions are only editable by admin users.
- Public and private toggles have been added. See #1278 for a breakdown of what this toggle does
- A model spec for conditions and goals has been added. Because this spec does not require tracking additional metadata, there are only two associations. A model specification is a recommended set of goals/conditions and their versions for all plans using the model. Unlike in constraints, plan scheduling specs are not prepopulated due to being unable to know what a plan's model is. This can be resolved after #1311 is completed.
- Scheduling Requests now record additional scheduling spec data in order to mirror Simulation Requests. Additionally, `analysis_id` has been officially made the pkey. This column has been a unique identifier for a while but was not officially changed until now. 
- The unusable "InMemoryStore" configuration has been removed. This code is completely unsupported, and the Scheduler uses a separate `Mock` implementation for tests.
- The Scheduling Services have been simplified. Specifically, `SpecificationService` and `SchedulerService` have been flattened to their single real implementation, which defers to the implementation of their repositories. Additionally, some tweaks have been made to the `PostgresRepository` implementation to remove some unneeded calls to the DB.

## Verification

DB Tests and Scheduling E2E Tests were updated to reflect the schema changes and for simplification.
SchedulingTests for versioning and enabling/disabling goals were added 

## Documentation

Docs ticket to be created

## Future work

- After #1311 is completed, add a prepopulate trigger for plan specs.

### CODE REVIEW

File: scheduler-server/sql/scheduler/tables/scheduling_condition_definition.sql [hunk 1]
```diff
@@ -0,0 +1,51 @@
+create table scheduling_condition_definition(
+  condition_id integer not null,
+  revision integer not null default 0,
+  definition text not null,
+  author text,
+  created_at timestamptz not null default now(),
+
+  constraint scheduling_condition_definition_pkey
+    primary key (condition_id, revision),
+  constraint scheduling_condition_definition_goal_exists
```
Comment (Mythicaeda) [inline scheduler-server/sql/scheduler/tables/scheduling_condition_definition.sql]: remove goal

File: scheduler-server/sql/scheduler/tables/scheduling_specification_conditions.sql [hunk 2]
```diff
@@ -1,25 +1,64 @@
 create table scheduling_specification_conditions (
   specification_id integer not null,
   condition_id integer not null,
```
Comment (Mythicaeda) [inline scheduler-server/sql/scheduler/tables/scheduling_specification_conditions.sql:3]: give fk

File: scheduler-server/src/main/java/gov/nasa/jpl/aerie/scheduler/server/services/ScheduleAction.java [hunk 3]
```diff
@@ -55,12 +55,11 @@ record Complete(ScheduleResults results, long analysisId, Optional<Long> dataset
   public Response run(final SpecificationId specificationId, final HasuraAction.Session session)
   throws NoSuchSpecificationException, IOException
   {
-    //record the plan revision as of the scheduling request time (in case work commences much later eg in worker thread)
-    //TODO may also need to verify the model revision / other volatile metadata matches one from request
-    final var specificationRev = this.specificationService.getSpecificationRevisionData(specificationId);
+    //record the plan and specification revision as of the scheduling request time (in case work commences much later in worker thread)
+    final var specificationRev = new ScheduleRequest(specificationId, this.specificationService.getSpecificationRevisionData(specificationId));
```
Comment (Mythicaeda) [inline scheduler-server/src/main/java/gov/nasa/jpl/aerie/scheduler/server/services/ScheduleAction.java]: rename this variable

File: scheduler-worker/src/main/java/gov/nasa/jpl/aerie/scheduler/worker/services/SynchronousSchedulerAgent.java [hunk 4]
```diff
@@ -359,17 +358,18 @@ private long getMerlinPlanRev(final PlanId planId)
   {
     return merlinService.getPlanRevision(planId);
   }
+
   /**
    * confirms that specification revision still matches that expected by the scheduling request
```
Comment (Mythicaeda) [inline scheduler-worker/src/main/java/gov/nasa/jpl/aerie/scheduler/worker/services/SynchronousSchedulerAgent.java]: change comment to "is still relevant"

**Review by adrienmaillard — APPROVED** [review]
LGTM, thank you for the walkthough.

**Comment by Mythicaeda** [comment]
Change pushed adds `priority` to the model goal spec, as per discussion with @dandelany

**Comment by Mythicaeda** [comment]
Change pushed was metadata, specifically:

- There was an `analyses` relationship from Goals to the Goal Analysis table that was skipped when writing the new metadata for the split goal tables. This relationship has been restored.

**Comment by dandelany** [comment]
👍 finished testing the UI with @duranb - merging this and aerie-ui/1149 together.

**Pull request merged**